gateway/handlers/external_auth_test.go

```diff
@@ -0,0 +1,72 @@
+package handlers
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func Test_External_Auth_Wrapper_FailsInvalidAuth(t *testing.T) {
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusForbidden)
+	}))
+	defer s.Close()
+
+	next := func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusNotImplemented)
+	}
+
+	passBody := false
+	handler := MakeExternalAuthHandler(next, s.URL, passBody)
+
+	req := httptest.NewRequest(http.MethodGet, s.URL, nil)
+	rr := httptest.NewRecorder()
+	handler(rr, req)
+
+	if rr.Code == http.StatusOK {
+		t.Errorf("Status incorrect, did not want: %d, but got %d", http.StatusOK, rr.Code)
+	}
+}
+
+func Test_External_Auth_Wrapper_PassesValidAuth(t *testing.T) {
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer s.Close()
+
+	next := func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusNotImplemented)
+	}
+
+	passBody := false
+	handler := MakeExternalAuthHandler(next, s.URL, passBody)
+
+	req := httptest.NewRequest(http.MethodGet, s.URL, nil)
+	rr := httptest.NewRecorder()
+	handler(rr, req)
+	want := http.StatusNotImplemented
+	if rr.Code != want {
+		t.Errorf("Status incorrect, want: %d, but got %d", want, rr.Code)
+	}
+}
+
+func MakeExternalAuthHandler(next http.HandlerFunc, upstreamURL string, passBody bool) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, _ := http.NewRequest(http.MethodGet, upstreamURL, nil)
+
+		res, err := http.DefaultClient.Do(req)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		if res.Body != nil {
+			defer res.Body.Close()
+		}
+
+		if res.StatusCode == http.StatusOK {
```

Comment:
It can also be good to allow non-200 response from the auth request to be returned directly to the user.  There can be headers or content in the body that enable or are required for the auth flow. 

nginx does this only for the 401 adn 403 responses http://nginx.org/en/docs/http/ngx_http_auth_request_module.html

Traefik returns all non-200 responses from the auth request https://docs.traefik.io/configuration/entrypoints/#forward-authentication

Reply:
I agree, I think it would be OK to add it after the first batch of PRs with an additional set of tests for that. We have some convenient 2xx checking-code somewhere we can re-use.

Reply:
sure